core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java

```diff
@@ -79,19 +81,42 @@ public long startRecovery(IndexShard indexShard, DiscoveryNode sourceNode,
      * Resets the recovery and performs a recovery restart on the currently recovering index shard
      *
      * @see IndexShard#performRecoveryRestart()
+     * @return newly created RecoveryTarget
      */
-    public void resetRecovery(long id, ShardId shardId) throws IOException {
-        try (RecoveryRef ref = getRecoverySafe(id, shardId)) {
-            // instead of adding complicated state to RecoveryTarget we just flip the
-            // target instance when we reset a recovery, that way we have only one cleanup
-            // path on the RecoveryTarget and are always within the bounds of ref-counting
-            // which is important since we verify files are on disk etc. after we have written them etc.
-            RecoveryTarget status = ref.status();
-            RecoveryTarget resetRecovery = status.resetRecovery();
-            if (onGoingRecoveries.replace(id, status, resetRecovery) == false) {
-                resetRecovery.cancel("replace failed"); // this is important otherwise we leak a reference to the store
-                throw new IllegalStateException("failed to replace recovery target");
+    public RecoveryTarget resetRecovery(final RecoveryRef recoveryRef, TimeValue retryCleanupTimeout,
+                                        StartRecoveryRequest currentRequest) {
+        long id = recoveryRef.status().recoveryId();
+        ShardId shardId = recoveryRef.status().shardId();
+
+        RecoveryTarget oldRecoveryTarget = null;
+        RecoveryTarget newRecoveryTarget = null;
+
+        try {
+            synchronized (onGoingRecoveries) {
+                // swap recovery targets in a synchronized block to ensure that the newly added recovery target is picked up by
+                // cancelRecoveriesForShard whenever the old recovery target is picked up
+                oldRecoveryTarget = onGoingRecoveries.remove(id);
+                if (oldRecoveryTarget == null) {
+                    return null;
+                }
+                assert recoveryRef.status() == oldRecoveryTarget : "recovery target not the one we were expecting";
+                assert oldRecoveryTarget.shardId().equals(shardId) :
+                    "shard id mismatch, expected " + shardId + " but was " + oldRecoveryTarget.shardId();
+
+                newRecoveryTarget = oldRecoveryTarget.retryCopy();
+                RecoveryTarget existingTarget = onGoingRecoveries.putIfAbsent(newRecoveryTarget.recoveryId(), newRecoveryTarget);
+                assert existingTarget == null : "found two RecoveryTarget instances with the same id";
             }
+
+            recoveryRef.close(); // close the reference held by RecoveryRunner
+            oldRecoveryTarget.resetRecovery(retryCleanupTimeout); // Closes the current recovery target
```

Comment:
we talked about an option to run this under the cancellable threads of the new target,  so we rely on it's monitor to cancel this if it takes too long. This will save us `retryCleanupTimeout` setting.